### PR TITLE
GEOMESA-2261 Kafka - concurrent performance improvements

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Added datastore option for changing geometry index resolution
* Replaced caffeine cache with ConcurrentHashMap and executor for expiry
* Modified SpatialIndex trait to reflect our usage, instead of QuadTree API
* GeoCQEngine updated to implement SpatialIndex
* Deprecated cache cleanup and consistency options as OBE
* Added optimized KafkaFeatureCache for point types to avoid creating centroids

Signed-off-by: Emilio Lahr-Vivaz <elahrvivaz@ccri.com>

### CODE REVIEW

**Comment by ben-manes** [comment]
Hi, can you please let me know if there is any improvements desired in Caffeine wrt
```
Replaced caffeine cache with ConcurrentHashMap and executor for expiry
```

When we move to JDK9+ in our base, we'll add eager expiration support via `CompletableFuture.delayedExecutor` to use the JVM-wide `ScheduledExecutorService`. Then you'll be able to opt into more proactive expiration, rather than waiting until activity occurs on the cache. Until then, you can call `cleanUp` periodically or use your own expiration handling like your PR does.

**Comment by elahrvivaz** [comment]
@ben-manes never mind, I see there's already a lot of documentation in the Caffeine docs.

**Comment by ben-manes** [comment]
> Do you have any idea of raw concurrent performance of a cache vs a ConcurrentHashMap for just puts and updates?

For updates, this [benchmark](https://github.com/ben-manes/caffeine/wiki/Benchmarks#write-100-1) shows the cache adds about 10% overhead. That's because contention on the entry is always going to dominate under that workload. I use a Zipfian distribution in benchmarks to create realistic workloads and their contention, since usually some entries are hotter than others, so both would be faster under a uniform distribution like your case.

For creation, its harder since that triggers an eviction if over capacity. I do have a benchmark but primarily to debug any degradations (e.g. Ehcache 3 used to degrade to a few evictions per second due to O(n) sampling). Its harder to give a realistic answer, but tends not to be a problem.

> We typically refresh the entire data set (~15k elements) every 10 seconds, and update both the cache and the spatial index.

Have you considered using Guava's `Suppliers.memoizeWithExpiration` instead? Then you can expire an entire map in one shot, rather than have per-entry policies. However that's a hard expiration rather than refresh. But if you know that all entries should be reloaded, a bulk query is faster than individual ones.

**Comment by elahrvivaz** [comment]
Thanks, I will look into `Suppliers.memoizeWithExpiration`. We use Caffeine caches in many other places, it's a really great library in general.

**Comment by elahrvivaz** [comment]
I don't think that the KDS is a performance bottleneck, but this should improve things slightly anyway, based on some informal testing.

**Comment by jahhulbert-ccri** [comment]
I think that we should likely get some CPU profiling on a geoserver in jboss with a super high load...

**Comment by elahrvivaz** [comment]
@jnh5y and I did some profiling of geoserver, and almost all of the CPU time was spent in rendering the map. There were some complex SLD functions that were slowing things down, but those aren't actually used anywhere except our test instance.
@jnh5y it didn't seem like there were any other obvious bottlenecks, right?

**Comment by jnh5y** [comment]
@elahrvivaz I am not sure what kind of load we were putting on the GeoServer in terms of writing to Kafka.

**Comment by elahrvivaz** [comment]
Yeah, we did not have a continuous ingest going, but we had intermittent (60s) updates I believe.

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/data/KafkaDataStore.scala [hunk 1]
```diff
@@ -287,10 +291,10 @@ object KafkaDataStore extends LazyLogging {
                                   consumerConfig: Properties,
                                   consumeFromBeginning: Boolean,
                                   cacheExpiry: Duration,
-                                  cacheCleanup: Duration,
-                                  cacheConsistency: Duration,
-                                  ticker: Ticker,
+                                  indexResolutionX: Int,
```
[the remark this answers is not in the document]
Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/data/KafkaDataStore.scala]: I updated the docs in the final PR, here: https://github.com/locationtech/geomesa/pull/1956/files#diff-6a53cd5b65bc16dc36050c20c3913ccfR51

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/KafkaFeatureCache.scala [hunk 2]
```diff
@@ -26,81 +27,71 @@ trait KafkaFeatureCache extends Closeable {
...
       *
       * TODO: https://geomesa.atlassian.net/browse/GEOMESA-1409
       */
     override def put(feature: SimpleFeature): Unit = {
-      val id = feature.getID
-      val old = cache.getIfPresent(id)
+      val pt = feature.getAttribute(geometryIndex).asInstanceOf[Point]
+      val featureState = new FeatureState(pt.getX, pt.getY, feature.getID)
+      logger.trace(s"${featureState.id} inserting with expiry $expiry")
+      val old = state.put(featureState.id, featureState)
```
[the remark this answers is not in the document]
Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/KafkaFeatureCache.scala]: we don't have any scala style checks - probably something we should add. In this case, I'd say that creating an option would be unnecessary overhead, as we're using a java collection and the potentially null value does not escape the function.

File: geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/KafkaFeatureCache.scala [hunk 3]
```diff
@@ -109,64 +100,85 @@ object KafkaFeatureCache {
...
+    * @param sft simple feature type
+    * @param support spatial index
+    * @param expiry expiry
+    */
+  class KafkaFeatureCacheCentroids(sft: SimpleFeatureType, support: SpatialIndexSupport, expiry: Duration)
+      extends KafkaFeatureCachePoints(sft, support, expiry) {
+
+    import org.locationtech.geomesa.utils.geotools.Conversions.RichGeometry
+
+    override def put(feature: SimpleFeature): Unit = {
```
[the remark this answers is not in the document]
Comment (elahrvivaz) [reply geomesa-kafka/geomesa-kafka-datastore/src/main/scala/org/locationtech/geomesa/kafka/index/KafkaFeatureCache.scala]: yeah, i just didn't copy the scaladocs. I'll update.

File: geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/BucketIndex.scala [hunk 4]
```diff
@@ -8,70 +8,89 @@
...
+    val j = snapY(y)
+    buckets(i)(j).put(key, item)
   }
 
-  override def remove(envelope: Envelope, item: T): Boolean = {
-    val (i, j) = getBucket(envelope)
-    buckets(i)(j).remove(item)
+  override def insert(envelope: Envelope, key: String, item: T): Unit = {
+    if (envelope.getArea > 0) {
+      logger.warn(s"This index only supports point inserts, but received $envelope")
```
[the remark this answers is not in the document]
Comment (elahrvivaz) [reply geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/index/BucketIndex.scala]: good point, I'll update

**Pull request merged**